### ISSUE TITLE
Open3d static reference (Update: Issue solved with tutorial for dynamic version)

### ISSUE THREAD

**Issue by petrasvestartas** [intel-isl/Open3D#2717]
Hi,

For static linking which header files I should reference for empty c++ console application project for windows visual studio @yxlao ?

I would like to learn how to link Open3D without CMakeLists using Visual Studio Windows.

Before I built project like this, but this does not explain me linking process:
```
cd docs\_static\C++
mkdir build
cd build
cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_INSTALL_PREFIX="C:\Install\Path" ..
cmake --build .


```

Using CMakeLists: https://github.com/intel-isl/Open3D/blob/master/docs/_static/C%2B%2B/CMakeLists.txt

**Comment by yxlao** [comment]
After running this in the terminal,
```
cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_INSTALL_PREFIX="C:\Install\Path" ..
```
there should be a Visual Studio `.sln` file. You could either open the solution file with Visual Studio GUI and go from there, or continue to use `cmake --build` in the terminal.

**Comment by petrasvestartas** [comment]
Dear @yxlao ,

This step works and in release folder I get three files:
1) TestVisualizer.exe
2) TestVisualizer.exp
3) TestVisualizer.lib

But what I want to do is to open a new Visual Studio project and reference the Open3D library and access function from there?
I will try to explain step what I did:

A) Compile Statically through CMake - Works:
![Qustion1](https://user-images.githubusercontent.com/18013985/101479965-4d7c6980-3953-11eb-8e1f-56599a1f67c7.png)

B) Build Visual Studio Project - Works:
![Qustion2](https://user-images.githubusercontent.com/18013985/101480012-5d944900-3953-11eb-9076-43b73a563c6d.png)


![Question3](https://user-images.githubusercontent.com/18013985/101480024-638a2a00-3953-11eb-9366-478267d3331f.png)

What I want to do is to ask a help from you is:

A) Create empty C++ Project File 
![CreateProject](https://user-images.githubusercontent.com/18013985/101480592-2ffbcf80-3954-11eb-9857-a4cbf87aef9c.png)


B) Link manually Open3D. This I do not know how to do this. Which Directory I should use to link Open3D header
![CreateProject1](https://user-images.githubusercontent.com/18013985/101480629-3c802800-3954-11eb-9d55-1c7d89c91b35.png)

**Comment by yxlao** [comment]
As I understand, linking with Open3D *with* CMake works fine, and now your target is to link Open3D *without* CMake. 

Here are the steps for reference:

1. When running cmake open3d config:
  - Specify the install dir by `-DCMAKE_INSTALL_PREFIX=C:\xxx\xxx`. You may choose a user directory to avoid having to use the Administrator permission.
  - You can use `-DBUILD_SHARED_LIBS=ON`, in this way there will be a single Open3D shared library
2. Build the `INSTALL` target, e.g. `cmake --build . --target INSTALL`
3. Now go to the `CMAKE_INSTALL_PREFIX` directory, you should see the include and lib directories to use

**Comment by petrasvestartas** [comment]
Dear @yxlao ,

Thank you for your patience with me, I am trying to learn.
I have very little knowledge with CMake.

I would lik to ask question for the first point you mentioned.

> Specify the install dir by -DCMAKE_INSTALL_PREFIX=C:\xxx\xxx. You may choose a user directory to avoid having to use the Administrator permission.

I tried to place -DCMAKE_INSTALL_PREFIX=C:\libs\Open3D line in CMakeLists.txt but it does not work:
![Prefix](https://user-images.githubusercontent.com/18013985/101613727-65fc8a80-3a0c-11eb-8c66-d21c636883d1.png)

Alo how and where should I place` -DBUILD_SHARED_LIBS=ON` ?

I attach the CMakeLists.txt, could you add these lines in .txt file so that I could understand them more clearly?
[CMakeLists.txt](https://github.com/intel-isl/Open3D/files/5664948/CMakeLists.txt)

**Comment by yxlao** [comment]
Something like this:

```ps1
# Build Open3D
mkdir build
cd build
cmake -G "Visual Studio 16 2019" -A x64 -DBUILD_SHARED_LIBS=ON -DCMAKE_INSTALL_PREFIX=C:\libs\Open3D ..
cmake --build . --target INSTALL
cd ..

# Build a project that links to Open3D
cd docs\_static\C++
cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_INSTALL_PREFIX=C:\libs\Open3D ..
cmake --build .
```

**Comment by petrasvestartas** [comment]
In then this is what worked without CMake:

The project below is compiled dynamically.
Then in visual studio I had to change 3 things:
![image](https://user-images.githubusercontent.com/18013985/101786135-e0590780-3afd-11eb-8021-b73dfa0b9b5c.png)
![image](https://user-images.githubusercontent.com/18013985/101786155-e51dbb80-3afd-11eb-8bb1-7d3aae42e87b.png)
![image](https://user-images.githubusercontent.com/18013985/101786192-efd85080-3afd-11eb-8231-a2bc1a5ba5ee.png)

In CMake I had to change these flags:
![Questi](https://user-images.githubusercontent.com/18013985/101786412-329a2880-3afe-11eb-85e9-ffa3093cafc1.png)

And the compilation + installation looks like this (it seems it is not possible to compile debug version, only release):

git clone --recursive https://github.com/intel-isl/Open3D
mkdir build
cd build

cmake -G "Visual Studio 16 2019" -A x64 -DCMAKE_INSTALL_PREFIX="C:\libs\Open3DDynamic\Open3D" ..
cmake --build . --config Release --target ALL_BUILD
cmake --build . --config Release --target INSTALL

**Comment by Jane-Tran** [comment]
@petrasvestartas I heple me, I do your tutorial, but not foud folder  Open3D_DIR

![ảnh](https://user-images.githubusercontent.com/44759245/103865833-5f3e6180-50f7-11eb-9faf-7027d00255fc.png)

**Comment by petrasvestartas** [comment]
The easiest way for Windows is to download precompiled version that is available in the most recent version 15.1 in github release.

If you are using cmake take a look how I use the precompiled version in my project:

https://github.com/ibois-epfl/Cockroach/blob/Cockroach/CMakeLists.txt

If you are not using cmake just link open3d in Visual Studio.

**Comment by SamHadjes** [comment]
Thanks a lot for answering! Linking to the precompiled binaries was the first thing I tried.
But in that case I get linking errors too.

I placed the binaries under : `C:\libs\Open3D`
In the project properties sheet I added: 
- under **C/C++ -> General -> Additionnal Include directories** : `C:\libs\Open3D\include; C:\libs\Open3D\include\open3d\3rdparty;`
- under **Linker -> General -> Additionnal Library Directories** :  `C:\libs\Open3D\lib; %(AdditionalLibraryDirectories)`
- under **Linker -> Input -> Additionnal Dependencies** : `Open3D.lib; %(AdditionalDependencies)`

When doing that VS is able to do linting and recognize Open3D functions thanks to the include, but I get the errors you can see on the image when building the Draw.cpp example in Release x64.
![image](https://user-images.githubusercontent.com/48731907/177954389-7fa7f02f-214e-4573-905d-4ba7914fb0a3.png)


Is there some configuration I missed or some external dependency I'm not aware of?

Thanks for your help!

**Comment by petrasvestartas** [comment]
64 bit and release?

**Comment by SamHadjes** [comment]
Yes, I made sure that the properties sheet was set up for that configuration.

**Comment by petrasvestartas** [comment]
I think your linking is wrong mine:

* C/C++ -> General -> Additionnal Include directories :
C:\IBOIS57\_Code\Software\CPP\CMAKE\super_build\example_cockroach\build\install\open3d\include
C:\IBOIS57\_Code\Software\CPP\CMAKE\super_build\example_cockroach\build\install\open3d\include\open3d\3rdparty

* under Linker -> Input -> Additionnal Dependencies : Open3D.lib; %(AdditionalDependencies)
install\open3d\lib\Open3D.lib

I do know why you placed headers here
under Linker -> General -> Additionnal Library Directories : C:\libs\Open3D\lib; %(AdditionalLibraryDirectories)

When do you you this directory? For me it is empty.

**Comment by SamHadjes** [comment]
Hi @petrasvestartas , thanks for your answer. For the linking I did, it was based in the recommended linking proposed for OpenCV.
If I'm not mistaken, the field under `Linker -> General -> Additionnal Library Directories` corresponds to the directory where my Open3D.lib file is placed.
This is the case on my installation.

I tested too with your way of doing it by only putting the complete path to the Open3D.lib file in `Linker -> Input -> Additionnal Dependencies` : `C:\libs\Open3D\lib\Open3D.lib;`

But it does not change anything when compiling, I get the exact same 6 linking errors as with the previous way of doing it.

I don't really understand when `fmt` is called for its types or why it is linking wrongly. I Have no clue on how to solve that.
